internal/apps/unit-of-work/main_test.go

```diff
@@ -0,0 +1,169 @@
+	// time goes to running the app as v1 and the other half to run it as v2 with
+	// a different workload. Each version produces one profile every
+	// profilePeriod.
+	//
+	// | totalDuration                                            |
+	// | v1                          | v2                         |
+	// | profile 1 | profile 2 | ... | profile 1 | profile 2 | ...|
+	// ------------------------------------------> time
+
+	totalDuration := 70 * time.Second // default
+	if s := os.Getenv("DD_TEST_APPS_TOTAL_DURATION"); s != "" {
+		totalDuration, err = time.ParseDuration(s)
+		require.NoError(t, err)
+	}
+
+	profilePeriod := 10 * time.Second // enough for 3 profiles per version
+	if s := os.Getenv("DD_TEST_APPS_PROFILE_PERIOD"); s != "" {
+		profilePeriod, err = time.ParseDuration(s)
+		require.NoError(t, err)
+	}
+
+	rps := 5
+	if s := os.Getenv("DD_TEST_APPS_REQUESTS_PER_SECOND"); s != "" {
+		_, err := fmt.Sscan(s, &rps)
+		require.NoError(t, err)
+	}
+
+	ddTags := strings.TrimSpace(fmt.Sprintf(
+		"%s run_id:%d DD_TEST_APPS_REQUESTS_PER_SECOND:%d DD_TEST_APPS_PROFILE_PERIOD:%s DD_TEST_APPS_TOTAL_DURATION:%s",
+		os.Getenv("DD_TAGS"),
+		rand.Uint64(),
+		rps,
+		profilePeriod,
+		totalDuration,
+	))
+	log.Printf("Using DD_TAGS: %s", ddTags)
+
+	versions := []struct {
+		Version   string
+		Endpoints []string // each endpoint is requested at the rps rate
+	}{
+		{"v1", []string{"/foo", "/bar"}},
+		{"v2", []string{"/foo", "/bar", "/bar"}},
+	}
+	for _, version := range versions {
+		t.Run(version.Version, func(t *testing.T) {
+			app := App{
+				DDTags:         ddTags,
+				ProfilePeriod:  profilePeriod,
+				ServiceVersion: version.Version,
+			}
+			app.Start(t)
+			defer app.Stop(t)
+
+			ctx, cancel := context.WithTimeout(context.Background(), totalDuration/time.Duration(len(versions)))
+			defer cancel()
+			eg := errgroup.Group{}
+			ticker := time.NewTicker(time.Second / time.Duration(rps))
+			defer ticker.Stop()
+
+		loop:
+			for {
+				select {
+				case <-ticker.C:
+					for _, endpoint := range version.Endpoints {
+						url := "http://" + app.HostPort + endpoint
+						eg.Go(func() error {
+							req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+							if err != nil {
+								return err
+							}
+							res, err := http.DefaultClient.Do(req)
+							if err != nil {
+								return err
+							}
+							return res.Body.Close()
+						})
+					}
+				case <-ctx.Done():
+					break loop
+				}
+			}
+			if err := eg.Wait(); !errors.Is(err, context.DeadlineExceeded) {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+type App struct {
+	ServiceVersion string
+	ProfilePeriod  time.Duration
+	DDTags         string
+	HostPort       string
+
+	proc *exec.Cmd
+}
+
+// Run launches the test app.
+func (a *App) Start(t *testing.T) {
+	// Start app
+	if a.HostPort == "" {
+		a.HostPort = "localhost:8080"
+	}
+	cmd := fmt.Sprintf(
+		"go build && exec ./unit-of-work -http %s -version %s -period %s",
```

Comment:
If I recall correctly from a previous discussion, the reason we're running the app like this (rather than doing everything in the same process) is so that the profiles & traces we get from the test app are more realistic. Otherwise we'd see `testing` frames in the call stacks, and we'd also see the load generation in addition to the actual work we're interested in. Is that right? Perhaps this should be documented?

Reply:
Yes, will add a comment.